README.md
# Checkmarx Scan Service for SBS 
## Usage in action

    runs-on: [docker, no-db]
    steps:
        - name: Run Checkmarx Scan for SBS services
          uses: LexisNexis-Public-GHA/SBS.Enablement.checkmarx-scan-service@1.0.0
        with:
          cx_username: ${{ secrets.CHECKMARX_USERNAME }}
          cx_password: ${{ secrets.CHECKMARX_PASSWORD }}
          cx_risk_threshold: 101
          cx_project_name: Accuity - SBS - ProjectName (%branch_name%)
          gh_repo_name: ${{ github.repository }}
          gh_branch_name: ${{ github.ref }}
          gh_pem_file: ${{ secrets.GH_APP_SECRET }}

**Note:** *%branch_name%* is a variable that represents the projects current branch name.

## Setting up in SBS Repo

TODO: add setup instructions

RELEASE_NOTES.md
# 1.1.0 (2021-05-12)
* ENBL-4332: adding cx_project_name as a required variable

# 1.0.0 (2021-05-05)
* Initial Release

business/github.js
const fs    = require("fs");
const jwt   = require("jsonwebtoken");
const axios = require("axios");

class CheckmarxGitHubApp {
    constructor(privateKey){
        this.id = "113731";
        this.installationId = "16713846";
        this.privateKey = privateKey;
    }

    generateAccessTokenAsync = async () => {
        
        //let key = fs.readFileSync(this.privateKeyPath, "utf8");

        let payload = {
            iat: Math.floor(new Date().getTime() / 1000) - 60,
            exp: Math.floor(new Date().getTime() / 1000) + (10 * 60),
            iss: this.id
        };
        
        let signingOptions = {
            algorithm:  "RS256"
        };

        let token = jwt.sign(payload, this.privateKey, signingOptions);

        // now that we have the bearer token, we need to exchange that for an access token

        let url = `https://api.github.com/app/installations/${this.installationId}/access_tokens`;

        let config = {
            headers: {
                "Authorization": `Bearer ${token}`,
                "Content-Type": "application/json"
            }
        };

        return axios.post(url, null, config)
            .then((result) => {
                return result.data;
            })
            .catch((err) => {
                console.log(err);
                return null;
            });
    };

}

module.exports = {
    CheckmarxGitHubApp : CheckmarxGitHubApp
}



dal/checkmarxapi.js
const axios     = require("axios");
const https     = require("https");
const cxApiUrl  = "https://checkmarx.rbxd.ds/cxrestapi";

const cmInstance = axios.create({
    httpsAgent: new https.Agent({
        rejectUnauthorized: false
    })
});

async function LoginAsync(username, password){
    let url = `${cxApiUrl}/auth/identity/connect/token`;

    let config = {
        headers: {
            "Content-Type" : "application/x-www-form-urlencoded"
        }
    }

    let params = new URLSearchParams();
    params.append("username", username);
    params.append("password", password);
    params.append("grant_type", "password");
    params.append("scope", "sast_rest_api");
    params.append("client_id", "resource_owner_client");
    params.append("client_secret", "014DF517-39D1-4453-B7B3-9930C563627C");

    return cmInstance.post(url, params, config)
        .then((result) => {
            return result.data.access_token;
        });
}

async function GetProjectByNameAndIdAsync(projectName, teamId, bearerToken){
    let url = `${cxApiUrl}/projects?projectName=${projectName}&teamId=${teamId}`;

    let config = {
        headers: {
            "Content-Type" : "application/x-www-form-urlencoded",
            "Authorization" : `Bearer ${bearerToken}`
        }
    };
    return cmInstance.get(url, config)
        .then((result) => {
            return result.data;
        })
        .catch((err) => {
            return null;
        });
}

async function CreateNewProectAsync(projectName, teamId, bearerToken){
    let url = `${cxApiUrl}/projects`;
    let body = {
        name: projectName,
        owningTeam : teamId,
        isPublic: true
    };

    let config = {
        headers: {
            "Content-Type" : "application/json",
            "Authorization" : `Bearer ${bearerToken}`
        }
    };

    return cmInstance.post(url, body, config)
        .then((result) => {
            return result.data;
        })
        .catch ((err) => {
            return null;
        });
}

async function UpsertStandardScanSettingsAsync(projectId, bearerToken){
    let url = `${cxApiUrl}/sast/scansettings`;
    let body = {
        projectId: projectId,
        presetId: 100001, //standard for sbs
        engineConfigurationId: 1 // standard for sbs
    };

    let config = {
        headers: {
            "Content-Type" : "application/json",
            "Authorization" : `Bearer ${bearerToken}`
        }
    };

    return cmInstance.post(url, body, config)
        .then ((result) => {
            return result.data;
        })
        .catch((err) => {
            console.log (err);
            return null;
        });
}

async function UpsertGitScanSettingsAsync(githubUrl, projectId, branchName, bearerToken){
    let url = `${cxApiUrl}/projects/${projectId}/sourceCode/remoteSettings/git`;

    let body = {
        url: githubUrl,
        branch: branchName
    };

    let config = {
        headers: {
            "Authorization": `Bearer ${bearerToken}`
        }
    };

    return cmInstance.post(url, body, config)
        .then((result) => {
            
            return result.status === 204;

        })
        .catch((err) => {
            console.log(err);
            throw err;
        });

}

async function ScheduleScanAsync (projectId, isIncremental, forceScan, comment, bearerToken){
    let url = `${cxApiUrl}/sast/scans`;

    let body = {
        projectId: projectId,
        isIncremental: isIncremental,
        isPublic: true,
        forceScan: forceScan,
        comment: comment
    };

    let config = {
        headers : {
            Authorization: `Bearer ${bearerToken}`
        }
    };

    return cmInstance.post(url, body, config)
        .then((result) => {
            return result.data;
        });
}

async function CheckScanAsync(scanId, bearerToken){
    let url = `${cxApiUrl}/sast/scans/${scanId}`;

    let config = {
        headers: {
            Authorization: `Bearer ${bearerToken}`
        }
    };

    return cmInstance.get(url, config)
        .then((result) => {
            return result.data;
        });
}


module.exports = {
    LoginAsync                  : LoginAsync,
    GetProjectByNameAndIdAsync  : GetProjectByNameAndIdAsync,
    CreateNewProectAsync        : CreateNewProectAsync,
    UpsertStandardScanSettingsAsync : UpsertStandardScanSettingsAsync,
    UpsertGitScanSettingsAsync: UpsertGitScanSettingsAsync,
    ScheduleScanAsync : ScheduleScanAsync,
    CheckScanAsync: CheckScanAsync
}

business/checkmarx.js
const dal = require("../dal/checkmarxapi");

class Checkmarx {
    constructor (username, password){
        this.username = username;
        this.password = password;
        this.bearerToken = null;
    }

    login = async() => {
        if (!this.bearerToken){
            this.bearerToken = await dal.LoginAsync(this.username, this.password);
        }
    }

    getProject = async(projectName, teamId) => {
        return await dal.GetProjectByNameAndIdAsync(projectName, teamId, this.bearerToken);
    };

    createProject = async (projectName, teamId) => {
        let createResult = await dal.CreateNewProectAsync(projectName, teamId, this.bearerToken);
        let projectId = createResult.id;

        console.log(createResult);

        console.log(`Using Project ID: ${projectId}`);

        if (projectId) {
            let scanSettingsResult = await dal.UpsertStandardScanSettingsAsync(projectId, this.bearerToken);
            if (scanSettingsResult){

                console.log("Got Scan results!");

                return projectId;
            } else {

                console.log("no scan results!");
                return null;
            }
        }
    };

    upsertGithubScanSettings = async(githubUrl, projectId, branchName) => {
        return await dal.UpsertGitScanSettingsAsync(githubUrl, projectId, branchName, this.bearerToken)
    };

    scheduleScan = async(projectId, isIncremental, forceScan) => {
        return await dal.ScheduleScanAsync(projectId, isIncremental, forceScan, "testing node app", this.bearerToken );
    }

    checkScan = async (scanId) => {
        return await dal.CheckScanAsync(scanId, this.bearerToken);
    };    
}

module.exports = {
    Checkmarx : Checkmarx
}

main.js
const core          = require("@actions/core");
const { Checkmarx } = require("./business/checkmarx.js");
const { CheckmarxGitHubApp } = require ("./business/github.js");

// checkmarx params
const username      = core.getInput("cx_username");
const password      = core.getInput("cx_password");

const isIncremental = core.getInput("cx_is_incremental").toLowerCase()  === "true";
const forceScan     = core.getInput("cx_force_scan").toLowerCase()      === "true";
const waitForScan   = core.getInput("cx_wait_for_scan").toLowerCase()   === "true";
const teamId        = core.getInput("cx_team_id");
// mutable
let projectName     = core.getInput("cx_project_name");

const riskThreshold = parseInt(core.getInput("cx_risk_threshold"))  || 101;     // fails github action if can result scanRiskSeverity is greater than or equal to riskThreshold
                                                                                // 101 effectively nullifies it as 100 is the highest scanRiskSeverity is 100

// github params
const pemfile       = core.getInput("gh_pem_file");
const repoName      = core.getInput("gh_repo_name");
const branchName    = core.getInput("gh_branch_name");

// derived variables
let simpleBranchName    = branchName.split('/').pop(); 
projectName = projectName.replace("%branch_name%", simpleBranchName);

core.info(`==> Using ProjectName: ${projectName}`);

function getExecutionTime(startDate){
    return Math.floor((new Date() - startDate) / 1000); 
}

async function main(){

    core.info("Starting Checkmarx Scan for SBS Services Action");

    core.info("==> Validating Variables");
    // check variables
    let requiredVariables = [
        {name: "cx_username"    , value: username},
        {name: "cx_password"    , value: password},
        {name: "gh_pem_file"    , value: pemfile},
        {name: "gh_repo_name"   , value: repoName},
        {name: "gh_branch_name" , value: branchName}
    ];

    let hasValidationError = false;
    for(i = 0; i < requiredVariables.length; i++){
        let o = requiredVariables[i];
        if (!o.value){
            hasValidationError = true;
            core.error(`====> Validation Error: ${o.name} value is required`);
        }
    }

    if (hasValidationError){
        throw "Unable to continue, fix validation errors";
    }

    let cm = new Checkmarx(username, password);
    core.info("==> Logging into Checkmarx API");
    await cm.login()
        .catch((err) => {
            console.error("====> Unable to login to checkmarx API, check username and password");
            throw err;
        });
    
    core.info ("==> Checking if Project Exists");
    core.debug(`====> Checking: ProjectName: ${projectName}`);
    core.debug(`                TeamId     : ${teamId}`);
    let projects = await cm.getProject(projectName, teamId);
    let projectId = -1;
    if (projects 
            && projects.length > 0 
            && projects[0].name 
            && projects[0].name.toLowerCase() === projectName.toLowerCase()){
                core.info("==> Project already exists!");
        projectId = projects[0].id;
    } else {
        core.info("Project does not exist... need to create");
        projectId = await cm.createProject(projectName, teamId);
    }

    // gather info for updating project for new scan
    if (!projectId || projectId < 0){
        throw "project id is not valid, something went wrong with creating or getting the information"
    }

    core.debug(`====> Using Project ID: ${projectId}`);

    core.info(`==> Generating GITHUB Access Token (valid for 1 hour)`);
    let cmApp = new CheckmarxGitHubApp(pemfile);
    let tokenResponse = await cmApp.generateAccessTokenAsync();

    core.info(`====> Github token Generated Successfully`);

    core.info("==> Updating Scan settings for project with new token");

    let token = tokenResponse.token;
    let githubUrl = `https://no-user:${token}@github.com/${repoName}`

    await cm.upsertGithubScanSettings(githubUrl, projectId, branchName)
        .catch((err) => {
            core.error("Something went wrong with updating scan settings in checkmarx");
            throw err;
        })

    core.info("====> Scan settings successfully updated");
    core.info("==> Scheduling Scan");

    // schedule scan.
    let scanResult = await cm.scheduleScan(projectId, isIncremental, forceScan)
                            .catch ((err) => {
                                core.error("Something went wrong with scheduling the scan");
                                throw err;
                            });

    let scanId = scanResult.id;
    core.info(`====> Scan ${scanId} successfully scheduled`);

    if (waitForScan){
        core.info("==> Waiting for scan to finish");
    }

    let startDate = new Date();

    let okToScan = true;
    while(waitForScan && okToScan){
        // wait 30 seconds
        await new Promise(r => setTimeout(r, 30000));

        let status = await cm.checkScan(scanId)
                                .catch((err) => {
                                    core.error("something went wrong with checking the status of the scan");
                                    throw err;
                                });

        if (status.status.id === 7){ // finished
            okToScan = false;

            let extTime = getExecutionTime(startDate);

            core.info(`====> Scan Completed in ${extTime} seconds`);
            core.info(`====> Scan Risk Severity: ${status.scanRiskSeverity}`);

            if (riskThreshold <= parseInt(status.scanRiskSeverity)){
                core.info(`==> Failing check, risk threshold breached`);
                throw "Scan Severity is greater than set risk threshold";
            }
        } else {
            core.info(`====> Scan Checked after ${getExecutionTime(startDate)} seconds.  Current Status: ${status.status.name} `);
        }
    }
}

main().then(() => {
    core.info("Scan Completed Successfully")
})
.catch((err) => {
    core.setFailed(err);
});